frontend/eslint-plugin-streamlit-custom/no-hardcoded-theme-values.js

```diff
@@ -0,0 +1,67 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+module.exports = {
+  meta: {
+    name: "no-hardcoded-theme-values",
+    type: "error",
+    docs: {
+      description: "Disallow hardcoded theme values",
+      category: "Best Practices",
+      recommended: true,
+    },
+    fixable: "code",
+    schema: [],
+  },
+  create(context) {
+    // Properties to check for values not allowed byt the allowedValuesRegex. We check the whole set of CSS properties with the allowed values,
+    // instead of checking for each property individually based on what they allow (e.g. only check 'background-color' for 'transparent'), since
+    // we do not need more fine granular matching.
+    const cssPropertiesToCheck =
+      /^(.*color|width|height|margin.*|padding.*|lineHeight|border.*|.*radius)$/i
+
+    // Match and highlight as errors all values that either do not contain the word 'theme' or are not a CSS built-in value.
+    // We also allow the value 0 because we use it extensively in our codebase.
+    // Also allow %, vh, vw units because they are relative and relative values are okay from our theming perspective.
+    const allowedValuesRegex =
+      /^(?!.*theme)(?!('|")?(transparent|solid|initial|none|inherit|auto|fit-content|collapse|0|[0-9]+(%|vh|vw))( !important)?('|")?$).*$/i
+
+    return {
```

Comment:
I have some concerns about this being strict enough. Specifically, if I'm understanding this correctly, I think this will find JS objects just fine as exemplified in the unit tests, but I'm worried about other types of styling.

For example,

`<div style={{ backgroundColor: 'red', width: '100px' }} />`

or 

```
const Button = styled.button`
  background-color: red;
  border-radius: 5px;
`;
```

Is the hope that we will prevent that way of styling via code review, and via making all styling use the js object syntax, we only need to target that via this rule?

Reply:
Great points! The good thing is that `<div style={{ backgroundColor: 'red', width: '100px' }} />` is already covered, because the `{ backgroundColor: 'red', width: '100px' }` is an ObjectExpression in the AST. I have just double-checked it manually, but did not add a test to keep our unit tests lean and keep them plain js files. With the knowledge that those are ObjectExpressions, I think that should be fine.

For the second example:
```javascript
const Button = styled.button`
  background-color: red;
  border-radius: 5px;
`;
```

I originally left it out because we have almost no occurrences in our codebase for the template string syntax (I just could find one) and our guidance is to rather use the Object notation as the preferred way. Though, I have just extended the rule to catch those cases above with the additional hint that also rather the Object notation should be used 🙂  The current rule does not match this properly:
```javascript
const A = styled.div`
  width: ${props => props.about};
`
```
because the value is empty in the AST / has a different structure. I think it should be fine to leave out the `width: ${props => props.about};` case and don't invest too much time into this because when you pass this, you likely do it anyways to pass the `theme` and because we don't encourage this notation.
Does this sound good to you?

Reply:
Yeah awesome! Regarding the first point 🤦. I think I was thinking of `<div style="color: red" />` and then I double checked to see if you can even do that in React (and good news you can't).

Thanks!